File: Consumption-Tests/Shared/Objective-C/ViewController+Extensions.m
#import "ViewController+Extensions.h"

@implementation AuthRequestBuilder

- (NSMutableURLRequest *)requestForSocketID:(NSString *)socketID channel:(PusherChannel *)channel {
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL: [[NSURL alloc] initWithString:@"http://localhost:9292/pusher/auth"]];
    NSString *dataStr = [NSString stringWithFormat: @"socket_id=%@&channel_name=%@", socketID, [channel name]];
    NSData *data = [dataStr dataUsingEncoding:NSUTF8StringEncoding];
    request.HTTPBody = data;
    request.HTTPMethod = @"POST";
    return request;
}

- (NSURLRequest *)requestForSocketID:(NSString *)socketID channelName:(NSString *)channelName {
    NSURLRequest *request = [[NSURLRequest alloc] initWithURL:[[NSURL alloc] initWithString:@"http://localhost:9292/pusher/auth"]];
    NSMutableURLRequest *mutableRequest = [[NSMutableURLRequest alloc] initWithURL: [[NSURL alloc] initWithString:@"http://localhost:9292/pusher/auth"]];

    NSString *dataStr = [NSString stringWithFormat: @"socket_id=%@&channel_name=%@", socketID, channelName];
    NSData *data = [dataStr dataUsingEncoding:NSUTF8StringEncoding];
    mutableRequest.HTTPBody = data;
    mutableRequest.HTTPMethod = @"POST";

    request = [mutableRequest copy];

    return request;
}

@end

@implementation ViewController (Extensions)

- (Pusher *)makeAndLaunchPusher {
    
    OCAuthMethod *authMethod = [[OCAuthMethod alloc] initWithSecret:@"YOUR_APP_SECRET"];
    PusherClientOptions *options = [[PusherClientOptions alloc] initWithAuthMethod:authMethod];

    // Use this if you want to try out your auth Endpoint
//    OCAuthMethod *endpointAuthMethod = [[OCAuthMethod alloc] initWithAuthRequestBuilder:[[AuthRequestBuilder alloc] init]];
//    PusherClientOptions *optionsWithEndpoint = [[PusherClientOptions alloc] initWithAuthMethod:endpointAuthMethod];

    Pusher *pusher = [[Pusher alloc] initWithAppKey:@"YOUR_APP_KEY" options:options];
    pusher.connection.delegate = self;

    pusher.connection.userDataFetcher = ^PusherPresenceChannelMember* () {
        NSString *uuid = [[NSUUID UUID] UUIDString];
        return [[PusherPresenceChannelMember alloc] initWithUserId:uuid userInfo:nil];
    };

    // bind to all events globally
    [pusher bindWithEventCallback:^void (PusherEvent *event) {
        NSString *message = [NSString stringWithFormat: @"Event received '%@'", event.eventName];

        if (event.channelName) {
            message = [message stringByAppendingFormat: @" on channel '%@'", event.channelName];
        }
        if (event.userId) {
            message = [message stringByAppendingFormat: @" from user '%@'", event.userId];
        }
        if (event.data) {
            message = [message stringByAppendingFormat: @" with data '%@'", event.data];
        }

        NSLog(@"%@", message);
    }];

    [pusher connect];

    // subscribe to a public channel
    PusherChannel *myChannel = [pusher subscribeWithChannelName:@"my-channel"];

    // bind a callback to an event on that channel
    [myChannel bindWithEventName:@"my-event" eventCallback:^void (PusherEvent *event) {
        NSString *dataString = event.data;
        // convert string to data for decoding
        NSData *data = [dataString dataUsingEncoding:NSUTF8StringEncoding];

        NSError *error;
        // parse data as json
        NSDictionary *jsonObject = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&error];

        NSString *name = jsonObject[@"name"];
        NSString *message = jsonObject[@"message"];

        NSLog(@"%@ says %@", name, message);
    }];

    // callback for member added event
    void (^onMemberAdded)(PusherPresenceChannelMember*) = ^void (PusherPresenceChannelMember *member) {
        NSLog(@"member added: %@", member);
    };

    // callback for member removed event
    void (^onMemberRemoved)(PusherPresenceChannelMember*) = ^void (PusherPresenceChannelMember *member) {
        NSLog(@"member removed: %@", member);
    };

    // subscribe to a presence channel
    PusherPresenceChannel *presChanExplicit = [pusher subscribeToPresenceChannelWithChannelName:@"presence-explicit" onMemberAdded:onMemberAdded onMemberRemoved:onMemberRemoved];

    // bind a callback on the presence channel
    [presChanExplicit bindWithEventName:@"testing" eventCallback: ^void (PusherEvent *event) {
        NSLog(@"Data: %@", event.data);
    }];

    // trigger a client event
    [presChanExplicit triggerWithEventName:@"client-testing" data:@{ @"developers" : @"developers developers developers" }];
    
    return pusher;
}

- (void)changedConnectionStateFrom:(enum ConnectionState)old to:(enum ConnectionState)new_ {
    NSLog(@"Old connection: %d, new connection: %d", (int)old, (int)new_);
}

- (void)debugLogWithMessage:(NSString *)message {
    NSLog(@"%@", message);
}

- (void)subscribedToChannelWithName:(NSString *)name {
    NSLog(@"Subscribed to channel %@", name);

    if ([name isEqual: @"presence-test"]) {
        NSLog(@"%@", [(PusherPresenceChannel *)[self.client.connection.channels findWithName:@"presence-test"] members]);
    }
}

- (void)failedToSubscribeToChannelWithName:(NSString *)name response:(NSURLResponse *)response data:(NSString *)data error:(NSError *)error {
    NSLog(@"Failed to subscribe to channel %@", name);
}

- (void)receivedError:(PusherError *)error {
    NSNumber *code = error.codeOC;
    if(code) {
        NSLog(@"Received error: (%ld) %@", [code longValue], error.message);
    } else {
        NSLog(@"Received error: %@", error.message);
    }
}

@end

File: Consumption-Tests/Shared/Swift/ViewController+Extensions.swift
import Foundation

import PusherSwift

class AuthRequestBuilder: AuthRequestBuilderProtocol {
    func requestFor(socketID: String, channelName: String) -> URLRequest? {
        var request = URLRequest(url: URL(string: "http://localhost:9292/pusher/auth")!)
        request.httpMethod = "POST"
        request.httpBody = "socket_id=\(socketID)&channel_name=\(channelName)".data(using: String.Encoding.utf8)
        return request
    }
}

struct DebugConsoleMessage: Codable {
    let name: String
    let message: String
}

extension ViewController {

    func makeAndLaunchPusher() -> Pusher {

        // Only use your secret here for testing or if you're sure that there's  security risk
        let pusherOptions = PusherClientOptions(authMethod: .inline(secret: "YOUR_APP_SECRET"))

//        // Use this if you want to try out your auth endpoint
//        let pusherOptions = PusherClientOptions(
//            authMethod: AuthMethod.authRequestBuilder(authRequestBuilder: AuthRequestBuilder())
//        )
        let pusher = Pusher(key: "YOUR_APP_KEY", options: pusherOptions)

        pusher.delegate = self

        pusher.connect()

        // bind to all events globally
        _ = pusher.bind(eventCallback: { (event: PusherEvent) in
            var message = "Received event: '\(event.eventName)'"

            if let channel = event.channelName {
                message += " on channel '\(channel)'"
            }
            if let userId = event.userId {
                message += " from user '\(userId)'"
            }
            if let data = event.data {
                message += " with data '\(data)'"
            }

            print(message)
        })

        // subscribe to a channel
        let myChannel = pusher.subscribe("my-channel")

        // bind a callback to event "my-event" on that channel
        _ = myChannel.bind(eventName: "my-event", eventCallback: { (event: PusherEvent) in

            // convert the data string to type data for decoding
            guard let json: String = event.data,
                let jsonData: Data = json.data(using: .utf8)
            else {
                print("Could not convert JSON string to data")
                return
            }

            // decode the event data as json into a DebugConsoleMessage
            let decodedMessage = try? JSONDecoder().decode(DebugConsoleMessage.self, from: jsonData)
            guard let message = decodedMessage else {
                print("Could not decode message")
                return
            }

            print("\(message.name) says \(message.message)")
        })

        // callback for member added event
        let onMemberAdded = { (member: PusherPresenceChannelMember) in
            print(member)
        }

        // subscribe to a presence channel
        let chan = pusher.subscribe("presence-channel", onMemberAdded: onMemberAdded)

        // triggers a client event on that channel
        chan.trigger(eventName: "client-test", data: ["test": "some value"])

        return pusher
    }
}

extension ViewController: PusherDelegate {

    // PusherDelegate methods

    func changedConnectionState(from old: ConnectionState, to new: ConnectionState) {
        // print the old and new connection states
        print("old: \(old.stringValue()) -> new: \(new.stringValue())")
    }

    func subscribedToChannel(name: String) {
        print("Subscribed to \(name)")
    }

    func debugLog(message: String) {
        print(message)
    }

    func receivedError(error: PusherError) {
        if let code = error.code {
            print("Received error: (\(code)) \(error.message)")
        } else {
            print("Received error: \(error.message)")
        }
    }
}
